masonry/src/tree_arena.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO - Factor out into a separate crate.
+#![allow(dead_code)]
+
+struct TreeNode<Item> {
+    id: u64,
+    item: Item,
+    children: Vec<TreeNode<Item>>,
+}
+
+#[derive(Default)]
+pub struct TreeArena<Item> {
+    roots: Vec<TreeNode<Item>>,
+}
+
+pub struct TreeArenaToken<'a, Item> {
+    children: &'a Vec<TreeNode<Item>>,
+}
+
+impl<'a, Item> Clone for TreeArenaToken<'a, Item> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, Item> Copy for TreeArenaToken<'a, Item> {}
+
+pub struct TreeArenaTokenMut<'a, Item> {
+    children: &'a mut Vec<TreeNode<Item>>,
+}
+
+#[cfg(FALSE)]
+struct TreeArena<Item> {
+    items: HashMap<u64, UnsafeCell<Item>>,
+    parents: HashMap<u64, u64>,
```

Comment:
We is this not inline in `items`?

Reply:
For clarity, I guess. It's essentially pseudocode anyway.